File: Editorial/129_Sum_Root_to_Leaf_Numbers/Go/1_Iterative_Preorder_Traversal.go
type item struct {
    node *TreeNode
    sum  int
}

func sumNumbers(root *TreeNode) int {
    if root == nil {
        return 0
    }
    stack := []*item{{root, root.Val}}
    rootToLeaf := 0
    for len(stack) > 0 {
        n := len(stack) - 1
        i := stack[n]
        stack = stack[:n]
        if i.node.Left == nil && i.node.Right == nil {
            rootToLeaf += i.sum
        } else {
            cur := i.sum * 10
            if i.node.Left != nil {
                stack = append(stack, &item{i.node.Left, cur + i.node.Left.Val})
            }
            if i.node.Right != nil {
                stack = append(stack, &item{i.node.Right, cur + i.node.Right.Val})
            }
        }
    }
    return rootToLeaf
}

File: Editorial/129_Sum_Root_to_Leaf_Numbers/C/1_Iterative_Preorder_Traversal.c
typedef struct {
    struct TreeNode* node;
    int partialSum;
} StackNode;
int sumNumbers(struct TreeNode* root) {
    if (root == NULL) return 0;
    StackNode myStack[10000] = {0};
    int top = -1;
    myStack[++top] = (StackNode){root, root->val};
    int rootToLeaf = 0;
    while (top >= 0) {
        StackNode p = myStack[top--];
        root = p.node;
        int currNumber = p.partialSum;
        if (root->left == NULL && root->right == NULL)
            rootToLeaf += currNumber;
        else {
            if (root->right != NULL)
                myStack[++top] = (StackNode){
                    root->right, currNumber * 10 + root->right->val};
            if (root->left != NULL)
                myStack[++top] =
                    (StackNode){root->left, currNumber * 10 + root->left->val};
        }
    }
    return rootToLeaf;
}
